Add CMake package install and config

nikitalita opened the pull request:
This adds package install and configuration to the CMakeLists for the main `dap` target. This makes it much easier to package for package managers like `vcpkg` and `conan`. 

The implementation is consistent with how your dependencies, nlohmann-json and gtest, are implementing this.

Couple of possible issues:
- I added a version number to the package name. This is for `vcpkg` reasons, as it will complain if the package doesn't have the version. I set it to `1.46.0` to match the protocol version you implement, but I'm not sure if that would be ideal for this, as there might be breaking changes in this package's API and we wouldn't be able to increase the major or minor version while keeping this consistent with the protocol version.
- This is probably something that should be addressed in a later PR, but ideally, we'd also offer the option of using externally provided `json` and `gtest` libraries through `find_package()`, as it might cause conflicts with depending packages that require a different json/gtest version. However, this would preclude building `json` and tests with sanitizers. I'm not sure how to handle this; maybe just fail if we're using external dependencies and the sanitizer option is set?

ben-clayton commented:
Thank you for taking the time to file this. cppdap [once had install targets](https://github.com/google/cppdap/commit/51cf2951d0b27c0ca653c02c0bb433c7cc30d002), but:
* There's currently no stable contract on the API / ABI. The protocol can change wildly between versions, so I'd like to see semantic versioning enforced before encouraging people to use prebuilts.
* Nobody maintained this, and people using the packages became grumpy when stuff broke. I do not have time to maintain packages, so I'd like to feel confident that someone would maintain packages before accepting this.
* cppdap is tiny and is very quick to compile. It has one transitive dependency on one of two JSON libraries. I would strongly encourage people to build from source than depending on prebuilt packages.

So unfortunately, I'm not hugely in favour of supporting packages, given the little attention I can give the project and the lack of strong versioning and API contracts in place. I do really appreciate the effort put into this PR.

Thanks,
Ben

nikitalita commented:
Neither vcpkg nor conan provide prebuilt packages; they are portage package managers, i.e. they download and build from source before installing. I am definitely not interested in packaging this as a prebuilt package, for the reasons you've listed here and in the commit log.

There are many reasons why including this as a submodule and building it along with the main source would not be feasible for many projects. Chief among them IMO is build system incompatibilities; building this from source as a submodule in the project would basically require the project to use CMake as its build system. This is my primary motivation for wanting to package this: I want to switch the old hacky DAP library in a debugger that I'm currently maintaining to this, and our build systems are incompatible. I know that there are other C++ DAP debuggers that are in the same boat, so I wanted to make this an option for them as well.

I think I can address the rest of your concerns:

> * There's currently no stable contract on the API / ABI. The protocol can change wildly between versions, so I'd like to see semantic versioning enforced before encouraging people to use prebuilts.
I can change the "MAJOR_VERSION_COMPATIBLE" to "MINOR_VERSION_COMPATIBLE". This would cause `CMake` to throw an error when building if downstream dependents specify a specific version number in `find_package()` and a package with the same minor version was not found, would that be sufficient? If necessary, I can also change the ConfigVersion.cmake file to throw an error if `find_package()` was not specified with a version number.

> * cppdap is tiny and is very quick to compile. It has one transitive dependency on one of two JSON libraries. I would strongly encourage people to build from source than depending on prebuilt packages.
Using one of the two JSON libraries can be added as a `feature` in both conan and vcpkg. I can change the configuration file template such that it attempts to find either nlohmann or rapid depending on which is set.

> * Nobody maintained this, and people using the packages became grumpy when stuff broke. I do not have time to maintain packages, so I'd like to feel confident that someone would maintain packages before accepting this.
The maintenance of the ports in either vcpkg or conan would not be the responsibility of the maintainers of this repo. I am already planning on maintaining the portfiles there, as this is going to be a dependency of a debugger I am working on. This PR is mostly so that the maintainers here have control over how the CMake package is configured, but I can do it in a patch set for the port files if necessary.

ben-clayton commented:
Sorry for the slow reply. I've been on vacation.

Okay - your arguments are convincing. Let's see if we can land this.

There's a couple of things the bots are unhappy with:

1.  CMake on the windows bots are erroring with:
```
CMake Error at CMakeLists.txt:216 (install):
  install TARGETS given no ARCHIVE DESTINATION for static library target
  "cppdap".
```
2. The ubuntu bots require a rebase to pass (requires https://github.com/google/cppdap/commit/8a320f56dc3f2340139eb342f3dd548a65329027)
3. The license checker is unhappy with `cmake/Config.cmake.in` not having a license header. Please can you add one?

Thanks,
Ben

nikitalita commented:
Excellent, I am happy to hear that :)

> 1. CMake on the windows bots are erroring with:
> 
> ```
> CMake Error at CMakeLists.txt:216 (install):
>   install TARGETS given no ARCHIVE DESTINATION for static library target
>   "cppdap".
> ```

I had already fixed this on my fork, so I pushed this to the branch.

> 2. The ubuntu bots require a rebase to pass (requires [8a320f5](https://github.com/google/cppdap/commit/8a320f56dc3f2340139eb342f3dd548a65329027))

Rebased on top of current `main`.

> 3. The license checker is unhappy with `cmake/Config.cmake.in` not having a license header. Please can you add one?

Done, and verified that it appears in the generated config file. I also bumped the copyright date on CMakeLists.txt.


I have added on my fork the option to use external JSON packages, but I will put that in a separate PR after this lands.

ben-clayton commented on CMakeLists.txt:
```diff
@@ -183,6 +193,48 @@ cppdap_set_target_options(cppdap)
+install(
+	TARGETS                  "${CPPDAP_TARGET_NAME}"
+	EXPORT                   "${CPPDAP_TARGETS_EXPORT_NAME}"
+    ARCHIVE DESTINATION      "${CMAKE_INSTALL_LIBDIR}"
```
nit: mix of tabs and whitespace here.

ben-clayton commented on CMakeLists.txt:
```diff
@@ -183,6 +193,48 @@ cppdap_set_target_options(cppdap)
+    ${CPPDAP_CMAKE_VERSION_CONFIG_FILE} COMPATIBILITY SameMinorVersion
+)
+configure_package_config_file(
+  ${CPPDAP_CMAKE_CONFIG_TEMPLATE}
```
nit: inconsistent indentation here and above.

ben-clayton reviewed (commented):
Getting there!

Windows bots are now complaining about:

```
CMake Error at C:/Program [Files/CMake/share/cmake-3.10/Modules/WriteBasicConfigVersionFile.cmake:35](https://cs.corp.google.com/piper///depot/google3/Files/CMake/share/cmake-3.10/Modules/WriteBasicConfigVersionFile.cmake?l=35) (message):
  Bad COMPATIBILITY value used for WRITE_BASIC_CONFIG_VERSION_FILE():
  "SameMinorVersion"
Call Stack (most recent call first):
  C:/Program [Files/CMake/share/cmake-3.10/Modules/CMakePackageConfigHelpers.cmake:207](https://cs.corp.google.com/piper///depot/google3/Files/CMake/share/cmake-3.10/Modules/CMakePackageConfigHelpers.cmake?l=207) (write_basic_config_version_file)
  CMakeLists.txt:207 (write_basic_package_version_file)
```

nikitalita commented:
> 



> Getting there!
> 
> Windows bots are now complaining about:
> 
> ```
> CMake Error at C:/Program [Files/CMake/share/cmake-3.10/Modules/WriteBasicConfigVersionFile.cmake:35](https://cs.corp.google.com/piper///depot/google3/Files/CMake/share/cmake-3.10/Modules/WriteBasicConfigVersionFile.cmake?l=35) (message):
>   Bad COMPATIBILITY value used for WRITE_BASIC_CONFIG_VERSION_FILE():
>   "SameMinorVersion"
> Call Stack (most recent call first):
>   C:/Program [Files/CMake/share/cmake-3.10/Modules/CMakePackageConfigHelpers.cmake:207](https://cs.corp.google.com/piper///depot/google3/Files/CMake/share/cmake-3.10/Modules/CMakePackageConfigHelpers.cmake?l=207) (write_basic_config_version_file)
>   CMakeLists.txt:207 (write_basic_package_version_file)
> ```

It looks that this was only added in CMake 3.11, we’re one version off.

In either case, package versioning was only added in CMake 3.x, so we’d either have to make creating the version config conditional based on the CMake version, or bump up the minimum required version. I know that this is used as a dependency by Swiftshader and other internal Google projects, so I would ask you: what’s the newest CMake version we can require without stepping on any toes? Or are we stuck with 2.8 as a minimum?

ben-clayton commented:
It looks like there's no issue updating to a newer CMake. Yay!

I believe we just need to replace:

https://github.com/google/cppdap/blob/03e975779ef8cb035bec5956840852099e5a24ed/kokoro/windows/presubmit.bat#L20

with:

```batch
SET PATH=C:\python36;C:\Program Files\cmake-3.23.1-windows-x86_64\bin;%PATH%
```

nikitalita commented:
Excellent; I have fixed the white space and bumped the required CMake version to 3.11.

nikitalita commented:
I've edited the kokoro script's PATH now, give it another go
